plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Builtins.hs

```diff
@@ -0,0 +1,24 @@
+{-# LANGUAGE LambdaCase #-}
+module UntypedPlutusCore.Check.Builtins (checkBuiltins) where
+
+import UntypedPlutusCore.Core
+
+-- Want this to inline so that 'go' can be specialized to the predicate.
+-- Probably it will be, but just make sure GHC can see it.
+{-# INLINABLE checkBuiltins #-}
+-- TODO: enable checking universe tags too: how?
```

Comment:
I think I wasn't sure how to check the tag properly.

Reply:
I put in the hook for this but I'm not using it yet since I'm not 100% sure how to do it, and anyway we don't have any new builtin types yet.

Reply:
I mean, does doing literally the same as with `fun` except with `SomeTypeIn uni` work? It has an `Ord` instance and if you need some other instances, we can add them as well. I.e. instead of writing, say, `AddInteger`, you write, say, `SomeTypeIn DefaultUniInteger`.

Not confident it would work, but I don't see problems with this approach right now.

Reply:
Hmm, not sure that would cover polymorphic builtin types, would it? How would I say that `(list (list integer))` was okay? We'll need something more fancy than just a set, I think.

Reply:
Right, I didn't think of polymorphic builtins. Anyways, there's `HasUniApply` that allows you to take apart an intra-universe application and check each of its parts individually, regardless of its kind. So you'd have `DefaultUniProtoList` in your `Set` and check that each part is either an application (and then recurse) or a ground tag (and then check if it's in the set). It does make the `DefaultUniApply` constructor a very special one, but we have it from the beginning, so there's no problem here.

Makes sense or am I missing something again?

Reply:
Yes it makes sense. I'm just afraid of writing this function :sweat_smile:  Okay, I'll give it a shot.